oscore-cf/src/main/java/org/eclipse/californium/oscore/OptionJuggle.java

```diff
@@ -0,0 +1,330 @@
+		for (Option option : optionList) {
+			optionSet.addOption(option);
+		}
+		return optionSet;
+	}
+
+	/**
+	 * Returns a new OptionSet, result, which doesn't contain any e options
+	 * @param optionSet the options
+	 * @return a new optionSet which have had the non-special e options removed
+	 */
+	public static OptionSet discardEOptions(OptionSet optionSet){
+		OptionSet result = new OptionSet();
+		boolean discard = false;
+		
+		for(Option opt : optionSet.asSortedList()){
+			Iterator<Integer> it = allEOptions.iterator();
+			while(!discard && it.hasNext()){
+				int tmp = it.next().intValue() - opt.getNumber();
+				if(tmp == 0){
+					discard = true;
+				}
+			}
+			if(!discard){
+				result.addOption(opt);
+			}
+			discard = false;
+		}
+		return result;
+	}
+	
+	/**
+	 * Sets the fake code in the coap header and returns the real code.
+	 * @param request the request that receives its fake code.
+	 * @return realCode the real code.
+	 */
+	public static Request setFakeCodeRequest(Request request){
+		Code fakeCode;
+
+		if(!request.getOptions().hasObserve()){
+			fakeCode = Code.POST;
+		}else{
+			fakeCode = Code.FETCH;
+		}
+		
+		return requestWithNewCode(request, fakeCode);
+	}
+	
+	/**
+	 * Sets the Request's CoAP Code with realCode
+	 * @param request the request that receives its real code
+	 * @param realCode the real code
+	 */
+	public static Request setRealCodeRequest(Request request, Code realCode){
+		return requestWithNewCode(request, realCode);
+	}
+	
+	/**
+	 * Sets the fake code in the coap header and returns the real code.
+	 * @param response the response that receives its fake code.
+	 * @return realCode the real code.
+	 */
+	public static Response setFakeCodeResponse(Response response){
+		return responseWithNewCode(response, ResponseCode.CHANGED);
+	}
+	
+	/**
+	 * Sets the realCode for a response
+	 * @param response
+	 * @param realCode
+	 */
+	public static Response setRealCodeResponse(Response response, ResponseCode realCode){
+		return responseWithNewCode(response, realCode);
+	}
+	
+	/**
+	 * Change the CoAP Code of the request to code
+	 * @param request the Request having its CoAP Code changed
+	 * @param code the new CoAP Code
+	 */
+	private static Request requestWithNewCode(Request request, Code code){
+		OptionSet options = request.getOptions();
+		byte[] payload = request.getPayload();
+		Token token = request.getToken();
+		EndpointContext destinationContext = request.getDestinationContext();
+		EndpointContext sourceContext = request.getSourceContext();
+		List<MessageObserver> messageObservers = request.getMessageObservers();
+		int mid = request.getMID();
+		Type type = request.getType();
+		
+		Request newRequest = new Request(code);
+		
+		newRequest.setOptions(options);
+		newRequest.setPayload(payload);
+		newRequest.setToken(token);
+		newRequest.setDestinationContext(destinationContext);
+		newRequest.setSourceContext(sourceContext);
+		newRequest.addMessageObservers(messageObservers);
```

Comment:
I miss the request's user context. 
If moved to the decrypt/encrypt please don't forget to add this.

Reply:
Thanks.